src/IssueViz/Editor/IIssueSpanCalculator.cs

```diff
@@ -21,27 +21,57 @@
 using System;
 using Microsoft.VisualStudio.Text;
 using SonarLint.VisualStudio.Core.Analysis;
+using SonarLint.VisualStudio.Core.Helpers;
 
 namespace SonarLint.VisualStudio.IssueVisualization.Editor
 {
     public interface IIssueSpanCalculator
     {
         /// <summary>
-        /// Returns the text span corresponding to the supplied analysis issue location
+        /// Returns the text span corresponding to the supplied analysis issue location.
+        /// Returns null if the location line hash is different from the snapshot line hash
         /// </summary>
-        SnapshotSpan CalculateSpan(IAnalysisIssueLocation location, ITextSnapshot currentSnapshot);
+        SnapshotSpan? CalculateSpan(IAnalysisIssueLocation location, ITextSnapshot currentSnapshot);
     }
 
     public class IssueSpanCalculator : IIssueSpanCalculator
     {
-        public SnapshotSpan CalculateSpan(IAnalysisIssueLocation location, ITextSnapshot currentSnapshot)
+        private readonly IChecksumCalculator checksumCalculator;
+
+        public IssueSpanCalculator()
+            : this(new ChecksumCalculator())
+        {
+        }
+        
+        internal IssueSpanCalculator(IChecksumCalculator checksumCalculator)
+        {
+            this.checksumCalculator = checksumCalculator;
+        }
+
+        public SnapshotSpan? CalculateSpan(IAnalysisIssueLocation location, ITextSnapshot currentSnapshot)
         {
+            if (location.StartLine > currentSnapshot.LineCount)
+            {
+                // Race condition: the line reported in the diagnostic is beyond the end of the file, so presumably
+                // the file has been edited while the analysis was being executed
+                return null;
+            }
+
+            // A start line of zero means the issue is file-level i.e. not associated with a particular line
+            var isFileLevelIssue = location.StartLine == 0;
+
             // SonarLint issues line numbers are 1-based, spans lines are 0-based
+            var startLine = currentSnapshot.GetLineFromLineNumber(location.StartLine - 1);
```

Comment:
Calling `currentSnapshot.GetLineFromLineNumber(-1)` in VS throws an exception, so I expect this line to throw for file-level issues. Does it actually work in VS?

Related to this, are there tests for file level issues? You said that tests had been moved from `IIssueToIssueMarkerConverterTests`, but I can't work out which one is the equivalent of `Convert_FileLevelIssue_ReturnsFilterableIssue`, and I can't see any tests where the issue StartLine is zero.

Reply:
We always [calculated](https://github.com/SonarSource/sonarlint-visualstudio/blob/master/src/Integration.Vsix/SonarLintTagger/IssueToIssueMarkerConverter.cs#L60) the span, even for a zero start line. It's possible that it doesn't work, I've asked Abbas for a rule that can raise a file level issue and I will check it.

Regarding the test -- I forgot to commit one (CalculateSpan_FileLevelIssue_HashNotChecked) and I'll write another one that compares the expected positions (once I verify that it does work).